Chore: (Docs) Storybook 6.5 testing updates (#17967)
storybookjs/storybook: docs/writing-tests/test-runner.md

> ---

Reviewer: At what point does it make sense to format the README to fit the structure we want for docs, and embed its content for this page?

Because this content seems like it's 90% the same.

Reviewer 2: Actually we want to do the opposite. I spoke to @shilman about this. The idea is to move everything to docs and then we can have the Reamde point to the Storybook docs.

Reviewer: That works too! 💯

Author: Marking this as solved.

title: 'Test runner'
---

Storybook test runner turns all of your stories into executable tests. It is powered by [Jest](https://jestjs.io/) and [Playwright](https://playwright.dev/).

- For those [without a play function](../writing-stories/introduction.md): it verifies whether the story renders without any errors.

> - For those [with a play function](../writing-stories/play-function.md): it also checks for errors in the play function and all assertions passed.

Reviewer: ```suggestion
- For those [with a play function](../writing-stories/play-function.md): it also checks for errors in the play function and that all assertions passed.
```

These tests run in a live browser and can be executed via the [command line](#cli-options) or your [CI server](#set-up-ci-to-run-tests).

## Setup

> The test runner is a standalone, framework-agnostic addon that runs parallel to your Storybook. You will need to take some additional steps to set it up properly. Detailed below is our recommendation to configure and execute it.

Reviewer: > addon

Are we using that term more generally than what gets registered under `main.js`'s `addons`?

Perhaps "library" or "package" would be better?

Reviewer 2: Good catch. I initially had utility. Would that work better?

Author: Rephrased it to utility. Marking this as solved.

Run the following command to install the addon and the required dependencies.

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-install.yarn.js.mdx',
    'common/storybook-test-runner-install.npm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

Update your `package.json` scripts and enable the test runner.

```json
{
  "scripts": {
    "test-storybook": "test-storybook"
  }
}
```

Start your Storybook with:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'angular/storybook-run-dev.with-builder.js.mdx',
    'common/storybook-run-dev.npm.js.mdx',
    'common/storybook-run-dev.yarn.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

<div class="aside">

> 💡 Storybook's test runner requires a running Storybook instance to run all the existing tests.

Reviewer: Not strictly true. It can also run against a built Storybook: https://github.com/storybookjs/test-runner/#running-against-a-deployed-storybook

Reviewer 2: > 💡 Storybook's test runner requires either a locally running or published Storybook instance  to run all the existing tests.

Author: With the latest commits. I've updated based on the conversation I and @kylegach had earlier today. @winkerVSbecks let me know if you're ok with it as well.

> 💡 Storybook's test runner requires a running Storybook instance to run all the existing tests.

Reviewer 2: ```suggestion
💡 Storybook's test runner requires either a locally running or published Storybook instance to run all the existing tests.
```

</div>

Finally, open a new terminal window and run the test runner with:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-execute.yarn.js.mdx',
    'common/storybook-test-runner-execute.npm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

## Configure

Test runner offers zero-config support for Storybook. However, you can run `test-storybook --eject` for more fine-grained control. It generates a `test-runner-jest.config.js` file at the root of your project, which you can modify. Additionally, you can extend the generated configuration file and provide [testEnvironmentOptions](https://github.com/playwright-community/jest-playwright#configuration) as the test runner also uses [jest-playwright](https://github.com/playwright-community/jest-playwright) under the hood.

### CLI Options

> The test runner is powered by [Jest](https://jestjs.io/) and accepts all its [CLI options](https://jestjs.io/docs/cli) (e.g., `--watch`, `--maxWorkers`).

Reviewer: ```suggestion
The test runner is powered by [Jest](https://jestjs.io/) and accepts all of its [CLI options](https://jestjs.io/docs/cli) (e.g., `--watch`, `--maxWorkers`).
```

Last I knew, @yannbf had purposefully restricted it to only accept a subset of Jest options. That has probably changed and my understanding is mistaken, but calling it out, just in case.

Reviewer 2: I might be mistaken then. @yannbf could you list out which flags are supported?

Reviewer: If it is indeed limited, then I think that's likely the table immediately below this paragraph.

Author: Based on [this](https://github.com/storybookjs/test-runner/blob/next/src/util/getParsedCliOptions.ts) the test runner CLI is indeed limited to a subset of commands and to avoid confusion the wording here should be instead something like:

"
The test runner is powered by [Jest](https://jestjs.io/) and accepts a subset of its [CLI options](https://jestjs.io/docs/cli) (e.g., `--watch`, `--maxWorkers`).
"

Author: Updated this with the wording I've mentioned above in the latest commits. Before marking it as solved I would like to check if that's accurate, because what the Readme lists is not all of them but actually the subset documented.

Reviewer 3: Yup we limited because in case we switch runners we have less trouble to support a subset of options rather than all options. People can still pass whatever option they want if they eject the config file though. Also, if they want to use certain options, they can open an issue and we will support it

Author: @yannbf thanks for the follow-up on this! Appreciate it! 🙏

| `--eject`                       | Creates a local configuration file to override defaults of the test-runner <br/>`test-storybook --eject`                         |

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-execute-with-flags.yarn.js.mdx',
    'common/storybook-test-runner-execute-with-flags.npm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

### Run tests against a deployed Storybook

By default, the test runner assumes that you're running it against a locally served Storybook on port `6006`. If you want to define a target URL to run against deployed Storybooks, you can use the `--url` flag or set the `TARGET_URL` environment variable. For example:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-execute-with-url.yarn.js.mdx',
    'common/storybook-test-runner-execute-with-url.npm.js.mdx',
    'common/storybook-test-runner-execute-with-url.env-var.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

## Set up CI to run tests

You can also configure the test runner to run tests on a CI environment. Documented below are some recipes to help you get started.

> ### Run against deployed Storybooks

Reviewer: This (either this heading or the following paragraph) is a little misleading, I think. If the Storybook is deployed, you can just pass the URL directly, no need for the complex recipe. That recipe is only for the case where you build your Storybook in CI and _don't_ deploy to a URL. I think the [README](https://github.com/storybookjs/test-runner/#running-in-ci) has a more accurate structure for these sections.

Reviewer 2: Yea agreed, this section is misleading at the moment. Here's my initial draft: https://www.notion.so/chromatic-ui/Test-Runner-ddc1d47c3889491087d5cef1496f1c0c#1e3f97aed2e94d6fa654f8ac6ad9b182

The first option should be `1. Run against locally built Storybooks`. It's not testing a deployed storybook. It's building it on CI and testing that.

Author: With the latest commits. I've updated based on the conversation I and @kylegach had earlier today. @winkerVSbecks let me know if you're ok with it as well.

- Playwright navigates to the story's page.
- The `preRender` function is executed.
- The story is rendered, and any existing `play` functions are executed.
- The `postRender` function is executed.

### Stories.json mode

When testing a local Storybook, the test runner transforms your story files into tests. For a remote Storybook, it uses the Storybook's [stories.json](../configure/overview.md#feature-flags) file (a static index of all the stories) to run the tests. Suppose you run into a situation where the local and remote Storybooks appear out of sync, or you might not even have access to the code. In that case, the `stories.json` file is guaranteed to be the most accurate representation of the deployed Storybook you are testing. To test a local Storybook using this feature, use the `--stories-json` flag as follows:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-with-stories-json.yarn.js.mdx',
    'common/storybook-test-runner-with-stories-json.npm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

<div class="aside">

💡 The `stories.json` mode is not compatible with watch mode.

</div>

> To disable it, use the `--no-stories-json` flag:

Reviewer: Why would someone need to disable it? The README includes this:

> By default, the test runner will detect whether your Storybook URL is local or remote, and if it is remote, it will run in "stories.json mode" automatically. To disable it, you can pass the --no-stories-json flag

Reviewer 2: @yannbf you explained this to me but I forget.

Reviewer: Sorry, I should've been more clear. I think the README explains the "why" well, and this doc should do something similar.

Author: @kylegach to give you a bit more context on this, I'm aware of it and left it as is intentionally,  as the idea was to spin up a page that details all the existing feature flags as what's currently documented needs some additional work, to which the people I've talked with were on board with. The perfect example of this is in the notion doc that @winkerVSbecks shared with you in a previous comment. Once I have it materialized I would love your feedback on it.

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-disable-stories-json.yarn.js.mdx',
    'common/storybook-test-runner-disable-stories-json.npm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

---

## Troubleshooting

### The test runner seems flaky and keeps timing out

If your tests time out with the following message:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-error.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

It might be that Playwright couldn't handle testing the number of stories you have in your project. Perhaps you have a large number of stories, or your CI environment has a really low RAM configuration. In such cases, you should limit the number of workers that run in parallel by adjusting your command as follows:

```json
{
  "scripts": {

>     "test-storybook:ci": "concurrently -k -s first -n \"SB,TEST\" -c \"magenta,blue\" \"yarn build-storybook --quiet && npx http-server storybook-static --port 6006 --silent\" \"wait-on tcp:6006 && yarn test-storybook --maxWorkers=2\""

Reviewer: The `--maxWorkers=2` flag is the only relevant bit here, right? I think we should avoid showing this complex recipe, if we can help it.

Author: Snippet was updated accordingly. Thanks for this

---
title: 'Test runner'
---

> Storybook test runner turns all of your stories into executable tests. It is powered by [Jest](https://jestjs.io/) and [Playwright](https://playwright.dev/).

Reviewer 3: You might want to double check every place in this PR that mentions this. Some places only mention Playwright, others mention Jest + Playwright

Author: I've updated the docs accordingly! There was a scenario that required clarity on this. The rest is left as is to focus the documentation as mentioning either can be misread or interpreted incorrectly.

  ]}
/>

<!-- prettier-ignore-end -->

### What's the difference between Chromatic and Test runner?

The test-runner is a generic testing tool that can run locally or on CI and be configured or extended to run all kinds of tests.

[Chromatic](https://www.chromatic.com/) is a cloud-based service that runs [visual](./visual-testing.md) and [interaction tests](./interaction-testing.md) (and soon accessibility tests) without setting up the test runner. It also syncs with your git provider and manages access control for private projects.

However, you might want to pair the test runner and Chromatic in some cases.

- Use it locally and Chromatic on your CI.
- Use Chromatic for visual and interaction tests and run other custom tests using the test runner.

## Advanced configuration

### Test hook API (experimental)

The test-runner renders a story and executes its [play function](writing-stories/play-function.md) if one exists. However, certain behaviors are impossible to achieve via the play function, which executes in the browser. For example, if you want the test-runner to take visual snapshots for you, this is possible via Playwright/Jest but must be executed in Node.

> The test-runner exports test hooks that can be overridden globally to enable use cases like visual or DOM snapshots. These hooks give you access to the test lifecycle before and after the story is rendered.

Reviewer 3: Do you think adding emphasis in before+after helps?

```suggestion
The test-runner exports test hooks that can be overridden globally to enable use cases like visual or DOM snapshots. These hooks give you access to the test lifecycle _before_ and _after_ the story is rendered.
```

Author: I'm good with it.

</div>

To enable the hooks API, you'll need to add a new configuration file inside your Storybook directory and set them up as follows:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-hooks-example.js.mdx',
    'common/storybook-test-runner-hooks-example.ts.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

<div class="aside">

💡 Except for the `setup` function, all other functions run asynchronously. Both `preRender` and `postRender` functions include two additional arguments, a [Playwright page](https://playwright.dev/docs/pages) and a context object which contains the `id`, `title`, and the `name` of the story.

</div>

When the test-runner executes, your existing tests will go through the following lifecycle:

- The `setup` function is executed before all the tests run.
- The context object is generated containing the required information.
- Playwright navigates to the story's page.
- The `preRender` function is executed.
- The story is rendered, and any existing `play` functions are executed.
- The `postRender` function is executed.

### Stories.json mode

> The test-runner transforms your story files into tests when testing a local Storybook. For a remote Storybook, it uses the Storybook's [stories.json](../configure/overview.md#feature-flags) file (a static index of all the stories) to run the tests.

Reviewer 3: Is "remote Storybook" enough for users to understand what that means? Or might we need a more beginner friendly version of this?

Reviewer 3: Also, we made a change that detects remote URLs and automatically runs with stories.json mode

#### Why?

Suppose you run into a situation where the local and remote Storybooks appear out of sync, or you might not even have access to the code. In that case, the `stories.json` file is guaranteed to be the most accurate representation of the deployed Storybook you are testing. To test a local Storybook using this feature, use the `--stories-json` flag as follows:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-with-stories-json.yarn.js.mdx',
    'common/storybook-test-runner-with-stories-json.npm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

<div class="aside">

💡 The `stories.json` mode is not compatible with watch mode.

</div>

If you need to disable it, use the `--no-stories-json` flag:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-disable-stories-json.yarn.js.mdx',
    'common/storybook-test-runner-disable-stories-json.npm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

> #### How do I check if Storybook has a `stories.json` file?

Reviewer 3: ```suggestion
#### How do I check if my Storybook has a `stories.json` file?
```

> Stories.json mode requires a `stories.json` file. Open a browser window and navigate to your deployed Storybook instance (for example, `https://your-storybook-url-here.com/stories.json`). You'll see a JSON file that starts with a `"v": 3` key, immediately followed by another key called "stories", which contains a map of story IDs to JSON objects.

Reviewer 3: ```suggestion
Stories.json mode requires a `stories.json` file. Open a browser window and navigate to your deployed Storybook instance (for example, `https://your-storybook-url-here.com/stories.json`). You should see a JSON file that starts with a `"v": 3` key, immediately followed by another key called "stories", which contains a map of story IDs to JSON objects. If that is the case, your Storybook supports stories.json mode.
```

Reviewer 3: Also if you need to point people to `buildStoriesJson` or `storyStoreV7` configs: https://storybook.js.org/docs/react/configure/overview#feature-flags

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-with-stories-json.yarn.js.mdx',
    'common/storybook-test-runner-with-stories-json.npm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

<div class="aside">

💡 The `stories.json` mode is not compatible with watch mode.

</div>

If you need to disable it, use the `--no-stories-json` flag:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-test-runner-disable-stories-json.yarn.js.mdx',
    'common/storybook-test-runner-disable-stories-json.npm.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

#### How do I check if my Storybook has a `stories.json` file?

Stories.json mode requires a `stories.json` file. Open a browser window and navigate to your deployed Storybook instance (for example, `https://your-storybook-url-here.com/stories.json`). You should see a JSON file that starts with a `"v": 3` key, immediately followed by another key called "stories", which contains a map of story IDs to JSON objects. If that is the case, your Storybook supports [stories.json mode](../configure/overview.md#feature-flags).

---

## Troubleshooting

> 

Reviewer 3: Possibly adding a new section after 0.0.7 is released:

#### The error output in the CLI is too short

By default, the test runner truncates error outputs at 1000 characters, and you can check the full output directly in Storybook, in the browser. If you do want to change that limit, however, you can do so by setting the `DEBUG_PRINT_LIMIT` environment variable to a number of your choosing, for example, `DEBUG_PRINT_LIMIT=5000 yarn test-storybook`.

Author: Fixed with the latest commit. Marking this as solved. Thanks for the heads up
